projecttacoma/bulk-export-app
# TACOSTRAT-1526  _elements Parameter Selection

# Summary

This PR adds the functionality to select elements that the user wants to include in the `_elements` parameter of the bulk export request. There are two ways to add parameters. The first way is to select a type and then select a number of elements that you want to export on that resource type. The other way is to select and element that will apply to all resource types on the server.

[JIRA Issue](https://jira.mitre.org/browse/TACOSTRAT-1526)

## New behavior

The  elements tab of the parameters section of the query-builder page is now implemented. The page has two multiselect input fields. 

The bottom one is a standard input field that is similar to the one on the type tab of this page where the user can select any number of elements, add, remove, or remove all elements. It is searchable and the selected options disappear from the dropdown and are shown in the input box at the top. This input is used for element filters that the user wants to apply to all resource types that are exported. Adding `id` for example is equivalent to adding a `_element=id` to the query string of the bulk export request.

The top input is more complicated. The user first selected a type they want to add element fitlers to. When the user selects a type, a modal is opened that prompts them to select elements in a dropdown similar to the one described above. Each type as some mandatory elements that are specified by the fhir spec. These mandatory elements are sourced from the `fhir-spec-tools` repo. The mandatory elements for each type are displayed in a box inside their respective modal when the user opens it. They also appear as disabled fields in the dropdown with tooltips that tell the user why they are disabled.

When the user selects a type and then elements on that type to filter for, the value is displayed in the input box like `Condition: id, name`. The user can clear all by clicking the icon on the right, clear individual types by clicking the x next to the type, or edit active type element filters by searching the resource type name in the input and clicking on option in the dropdown.

All of the actions described above actively change the export request string shown at the top of the page. When the user executes the workflow to view the output, the element filtering feature should be actively reflected in the returned files.


## Code changes
- `.prettierignore` added the .github file from the PR to put the app on github pages.
- `data/allSupportedElements.ts` to get all supported elements from `fhir-spec-tools` to use in dropdown population
- `data/mandatoryElements.ts` file to get mandatory elements from `fhir-spec-tools`
- `data/supportedElements.ts` to get elements, type record to get elements associated with each type from `fhir-spec-tools`
- `data/supportedResources.ts` get the resource types available on the server.
- `package-lock.json` add `fhir-spec-tools`
- `package.json` added `fhir-spec-tools`
- `src/app/global.module.css` updated styles
- `src/app/layout.tsx` moved mantine provider to `Providers.tsx` to  make use of a global theme object to synchronize mantine component styles.
- `src/app/providers.tsx` added theme to make multiselects have the same properties as well as cards on the page to be the same.
- `src/app/query-builder/page.tsx` added the element page.
- `src/components/export-execution/requested-files.tsx` added loading indication to when files are being processed after fetched from server.
- `src/components/parameter-pages/elements-modal.tsx` modal component that shows the mandatory elements and the MultiSelect to select elements on each type.
- `src/components/parameter-pages/elements-params-page.tsx` display both multiselects on the elements tab in the query builder page
- `src/components/parameter-pages/type-params-page.tsx` got rid of unnecessary fragment wrapper.
- `src/components/query-string.tsx` updated to reflect changes in the `exportRequestBuilder`.
- `src/state/element-params-state.ts` new recoil state for element parameters that apply to all types
- `src/state/type-element-params-state.ts` new recoil state for element parameters that apply to specific types.
- `src/util/exportRequestBuilder.ts` updated to build _elements parameter
- `src/util/exportRequestBuilders.ts` changed the name

# Testing guidance
- run `npm run check`
- Check that prettier and eslint cause no errors
- `npm run build`
- `npm run start`

Test on single type
- Select a type from the type tab.
- Select some elements on that type in the elements tab in the top multiselect
- Go to the export execution page and check that only the types you selected are exported, as well as the mandatory elements specified in the modal window

Test all type _elements filter
- Select an element from the bottom muliselect on the elements tab that applies to multiple elements, like `status`.
- Check that this filter applys to all exported resource types

Test combinations
- Test with some of the first test and some of the second test and verify that it works as expected.
